crypto/ec/asm/ecp_sm2p256-riscv64.pl

```diff
@@ -0,0 +1,1251 @@
+	and $s5, $s5, $c0
+	and $s6, $s6, $c0
+	and $s7, $s7, $c0
+	or $t4, $t4, $s4
+	or $t5, $t5, $s5
+	or $t6, $t6, $s6
+	or $t7, $t7, $s7
+
+	// 3*a
+	add $t4, $t4, $t0
+	sltu $c0, $t4, $t0
+
+	add $t5, $t5, $t1
+	sltu $c1, $t5, $t1
+	add $t5, $t5, $c0
+	sltu $c0, $t5, $c0
+	add $c0, $c0, $c1
+
+	add $t6, $t6, $t2
+	sltu $c1, $t6, $t2
+	add $t6, $t6, $c0
+	sltu $c0, $t6, $c0
+	add $c0, $c0, $c1
+
+	add $t7, $t7, $t3
+	sltu $c1, $t7, $t3
+	add $t7, $t7, $c0
+	sltu $c0, $t7, $c0
+	add $c0, $c0, $c1
+
+	// Sub polynomial
+	sltu $c3, $t4, $s0
+ 	sub $s4, $t4, $s0
+
+	sltu $c1, $t5, $c3
+	sub $s5, $t5, $c3
+	sltu $c3, $s5, $s1
+	sub $s5, $s5, $s1
+	add $c3, $c3, $c1
+
+	sltu $c1, $t6, $c3
+	sub $s6, $t6, $c3
+	sltu $c3, $s6, $s2
+	sub $s6, $s6, $s2
+	add $c3, $c3, $c1
+
+	sltu $c1, $t7, $c3
+	sub $s7, $t7, $c3
+	sltu $c3, $s7, $s3
+	sub $s7, $s7, $s3
+	add $c3, $c3, $c1
+
+	slt $c0, $c0, $c3
+	negw $c1, $c0
+	addw $c0, $c0, -1
+
+	and $t4, $t4, $c1
+	and $t5, $t5, $c1
+	and $t6, $t6, $c1
+	and $t7, $t7, $c1
+	and $s4, $s4, $c0
+	and $s5, $s5, $c0
+	and $s6, $s6, $c0
+	and $s7, $s7, $c0
+	or $t4, $t4, $s4
+	or $t5, $t5, $s5
+	or $t6, $t6, $s6
+	or $t7, $t7, $s7
+
+	// Store results
+	sd $t4, 0($i0)
+	sd $t5, 8($i0)
+	sd $t6, 16($i0)
+	sd $t7, 24($i0)
+___
+$code.= load_regs();
+$code.=<<___;
+	ret
+.size ecp_sm2p256_mul_by_3,.-ecp_sm2p256_mul_by_3
+
+
+// void ecp_sm2p256_add(BN_ULONG *r,const BN_ULONG *a,const BN_ULONG *b);
+.globl	ecp_sm2p256_add
+.type	ecp_sm2p256_add,%function
+.p2align 5
+ecp_sm2p256_add:
+___
+    &bn_mod_add(".Lpoly");
+$code.=<<___;
+	ret
+.size ecp_sm2p256_add,.-ecp_sm2p256_add
+
+// void ecp_sm2p256_sub(BN_ULONG *r,const BN_ULONG *a,const BN_ULONG *b);
+.globl	ecp_sm2p256_sub
+.type	ecp_sm2p256_sub,%function
+.p2align 5
+ecp_sm2p256_sub:
+___
+    &bn_mod_sub(".Lpoly");
+$code.=<<___;
+	ret
+.size ecp_sm2p256_sub,.-ecp_sm2p256_sub
+
+// void ecp_sm2p256_sub_mod_ord(BN_ULONG *r,const BN_ULONG *a,const BN_ULONG *b);
+.globl	ecp_sm2p256_sub_mod_ord
+.type	ecp_sm2p256_sub_mod_ord,%function
+.p2align 5
+ecp_sm2p256_sub_mod_ord:
+___
+    &bn_mod_sub(".Lord");
+$code.=<<___;
+	ret
+.size ecp_sm2p256_sub_mod_ord,.-ecp_sm2p256_sub_mod_ord
+
+// void ecp_sm2p256_div_by_2(BN_ULONG *r,const BN_ULONG *a);
+.globl	ecp_sm2p256_div_by_2
+.type	ecp_sm2p256_div_by_2,%function
+.p2align 5
+ecp_sm2p256_div_by_2:
+___
+    &bn_mod_div_by_2(".Lpoly_div_2");
+$code.=<<___;
+	ret
+.size ecp_sm2p256_div_by_2,.-ecp_sm2p256_div_by_2
+
+// void ecp_sm2p256_div_by_2_mod_ord(BN_ULONG *r,const BN_ULONG *a);
+.globl	ecp_sm2p256_div_by_2_mod_ord
+.type	ecp_sm2p256_div_by_2_mod_ord,%function
+.p2align 5
+ecp_sm2p256_div_by_2_mod_ord:
+___
+    &bn_mod_div_by_2(".Lord_div_2");
+$code.=<<___;
+	ret
+.size ecp_sm2p256_div_by_2_mod_ord,.-ecp_sm2p256_div_by_2_mod_ord
+
+.macro RDC
+	// a = |  t7   | ... | t0  |, where ti are 64-bit quantitiet
+	//   = | s7| s6| ... |--|--|, where ai are 32-bit quantitiet
+	// |    t7     |    t6     |    t5     |    t4     |
```

Comment:
The remapping you have done here (from the arm code) makes this fairly hard to follow..